execution_engine/src/storage/trie/mod.rs

```diff
@@ -511,40 +510,97 @@ impl<K, V> Trie<K, V> {
     }
 }
 
-pub(crate) type LazyTrieLeaf<K, V> = Either<Bytes, Trie<K, V>>;
-
-pub(crate) fn lazy_trie_tag(bytes: &[u8]) -> Option<TrieTag> {
-    bytes.first().copied().and_then(TrieTag::from_u8)
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum LazilyDeserializedTrie {
+    Leaf(Bytes),
+    Node {
+        pointer_block: Box<PointerBlock>,
+    },
+    Extension {
+        /// Extension node affix bytes.
+        affix: Bytes,
+        /// Extension node pointer.
+        pointer: Pointer,
+    },
 }
 
-pub(crate) fn lazy_trie_deserialize<K, V>(
-    bytes: Bytes,
-) -> Result<LazyTrieLeaf<K, V>, bytesrepr::Error>
-where
-    K: FromBytes,
-    V: FromBytes,
-{
-    let trie_tag = lazy_trie_tag(&bytes);
+impl LazilyDeserializedTrie {
+    pub(crate) fn iter_children(&self) -> DescendantsIterator {
+        match self {
+            LazilyDeserializedTrie::Leaf(_) => {
+                // Leaf bytes does not have any children
+                DescendantsIterator::ZeroOrOne(None)
+            }
+            LazilyDeserializedTrie::Node { pointer_block } => DescendantsIterator::PointerBlock {
+                iter: pointer_block.0.iter().flatten(),
+            },
+            LazilyDeserializedTrie::Extension { pointer, .. } => {
+                DescendantsIterator::ZeroOrOne(Some(pointer.into_hash()))
+            }
+        }
+    }
 
-    if trie_tag == Some(TrieTag::Leaf) {
-        Ok(LazyTrieLeaf::Left(bytes))
-    } else {
-        let deserialized: Trie<K, V> = bytesrepr::deserialize(bytes.into())?;
-        Ok(LazyTrieLeaf::Right(deserialized))
+    pub(crate) fn try_deserialize_leaf_key<K: FromBytes>(
+        &self,
+    ) -> Result<(K, &[u8]), bytesrepr::Error> {
+        match self {
+            LazilyDeserializedTrie::Leaf(leaf_bytes) => {
+                let (tag_byte, rem) = u8::from_bytes(leaf_bytes)?;
+                let tag = TrieTag::from_u8(tag_byte).ok_or(bytesrepr::Error::Formatting)?;
+                if let TrieTag::Leaf = tag {
+                    K::from_bytes(rem)
+                } else {
+                    Err(bytesrepr::Error::Formatting)
+                }
+            }
+            LazilyDeserializedTrie::Node { .. } | LazilyDeserializedTrie::Extension { .. } => {
+                Err(bytesrepr::Error::Formatting)
```

Comment:
I'm thinking the `Formatting` error may be confusing here. Technically, the "Trie" deserialization was successful, so what do you think about this:

```suggestion
                Ok(None)
```

Reply:
Yes, makes more sense to return Ok(None) here, though we'll have to convert the error in the callers somehow. I'll try to make the change.

Reply:
Went into a slightly different direction since returning and Optional value here was leading to a lot of useless code. We are actually getting the key only from a leaf so we shouldn't even be ever looking at the other variants. Added a new type for the leaf bytes in this commit: https://github.com/casper-network/casper-node/pull/4083/commits/765677f7c07be92bd21cbfd3303dfcf9e74cd819